### PR TITLE
add support for image1d_buffer

### PR DESCRIPTION
**Pull request by rjodinchr**
This implementation has been tested using the OpenCL-CTS patch with https://github.com/KhronosGroup/OpenCL-CTS/pull/1806

### CODE REVIEW

**Comment by rjodinchr** [comment]
This PR depends on an update of clspv

File: src/api.cpp [hunk 1]
```diff
@@ -710,7 +710,9 @@ cl_int CLVK_API_CALL clGetDeviceInfo(cl_device_id dev,
         size_ret = sizeof(val_uint);
         break;
     case CL_DEVICE_IMAGE_MAX_BUFFER_SIZE:
-        val_sizet = device->vulkan_limits().maxImageDimension1D;
+        val_sizet =
+            std::min((uint64_t)device->vulkan_limits().maxTexelBufferElements,
+                     device->max_mem_alloc_size());
```
Comment (kpet) [inline src/api.cpp:715]: Let's push this into a `cvk_device::image_max_buffer_size` function.

File: src/api.cpp [hunk 2]
```diff
@@ -4809,29 +4841,37 @@ cl_int CLVK_API_CALL clGetSupportedImageFormats(cl_context context,
...
         if ((features & required_format_feature_flags) ==
             required_format_feature_flags) {
+            VkComponentMapping components_sampled, components_storage;
+            get_component_mappings_for_channel_order(clfmt.image_channel_order,
+                                                     &components_sampled,
+                                                     &components_storage);
+            if ((components_sampled != components_storage) &&
+                (image_type == CL_MEM_OBJECT_IMAGE1D_BUFFER)) {
+                continue;
+            }
```
Comment (kpet) [inline src/api.cpp:4863]: I'm a bit confused by this condition. Those swizzles are unused for 1D buffer images. Why is this needed?

File: src/api.cpp [hunk 3]
```diff
@@ -4965,10 +5005,23 @@ cl_int CLVK_API_CALL clEnqueueReadImage(
...
     }
 
-    return cvk_enqueue_image_copy(command_queue, CL_COMMAND_READ_IMAGE, image,
-                                  blocking_read, origin, region, row_pitch,
-                                  slice_pitch, ptr, num_events_in_wait_list,
-                                  event_wait_list, event);
+    auto internal_image = static_cast<cvk_image*>(image);
+    if (internal_image->is_backed_by_buffer_view()) {
+        auto cmd = new cvk_command_buffer_host_copy(
+            command_queue, CL_COMMAND_READ_BUFFER,
```
Comment (kpet) [inline src/api.cpp:5011]:
```suggestion
            command_queue, CL_COMMAND_READ_IMAGE,
```

Otherwise, the application could see the wrong command type.

File: src/api.cpp [hunk 4]
```diff
@@ -5028,11 +5081,24 @@ cl_int CLVK_API_CALL clEnqueueWriteImage(
...
+    if (internal_image->is_backed_by_buffer_view()) {
+        auto cmd = new cvk_command_buffer_host_copy(
+            command_queue, CL_COMMAND_WRITE_BUFFER,
+            static_cast<cvk_buffer*>(internal_image->buffer()), ptr,
+            origin[0] * internal_image->element_size(),
+            region[0] * internal_image->element_size());
+        auto err = command_queue->enqueue_command_with_deps(
+            cmd, blocking_write, num_events_in_wait_list, event_wait_list,
+            event);
+        return err;
```
Comment (kpet) [inline src/api.cpp:5094]: I think you probably want to push this code to `cvk_enqueue_image_copy` so it's the same for read image and write image commands.

File: src/api.cpp [hunk 5]
```diff
@@ -5102,18 +5168,49 @@ clEnqueueCopyImage(cl_command_queue cq, cl_mem src_image, cl_mem dst_image,
     if (!src_img->has_same_format(dst_img)) {
         return CL_IMAGE_FORMAT_MISMATCH;
     }
+    assert(src_img->element_size() == dst_img->element_size());
```
Comment (kpet) [inline src/api.cpp:5171]: This should not be an assert. OpenCL specifies the following error condition:
>CL_IMAGE_FORMAT_MISMATCH if src_image and dst_image do not use the same image format.

File: src/api.cpp [hunk 6]
```diff
@@ -5102,18 +5168,49 @@ clEnqueueCopyImage(cl_command_queue cq, cl_mem src_image, cl_mem dst_image,
...
     std::array<size_t, 3> dst_orig = {dst_origin[0], dst_origin[1],
                                       dst_origin[2]};
     std::array<size_t, 3> reg = {region[0], region[1], region[2]};
 
-    auto cmd = std::make_unique<cvk_command_image_image_copy>(
-        command_queue, src_img, dst_img, src_orig, dst_orig, reg);
+    if (src_img->is_backed_by_buffer_view() &&
+        dst_img->is_backed_by_buffer_view()) {
+        auto cmd = new cvk_command_copy_buffer(
+            command_queue, CL_COMMAND_COPY_BUFFER,
```
Comment (kpet) [inline src/api.cpp:5182]:
```suggestion
            command_queue, CL_COMMAND_COPY_IMAGE,
```

That's what the application should see.

File: src/api.cpp [hunk 7]
```diff
@@ -5102,18 +5168,49 @@ clEnqueueCopyImage(cl_command_queue cq, cl_mem src_image, cl_mem dst_image,
...
+            static_cast<cvk_buffer*>(dst_img->buffer()),
+            src_origin[0] * src_img->element_size(),
+            dst_origin[0] * dst_img->element_size(),
+            region[0] * src_img->element_size());
+
+        return command_queue->enqueue_command_with_deps(
+            cmd, num_events_in_wait_list, event_wait_list, event);
+    } else if (src_img->is_backed_by_buffer_view()) {
+        auto cmd = std::make_unique<cvk_command_buffer_image_copy>(
+            CL_COMMAND_COPY_BUFFER_TO_IMAGE, command_queue,
```
Comment (kpet) [inline src/api.cpp:5193]:
```suggestion
            CL_COMMAND_COPY_IMAGE, command_queue,
```

File: src/api.cpp [hunk 8]
```diff
@@ -5102,18 +5168,49 @@ clEnqueueCopyImage(cl_command_queue cq, cl_mem src_image, cl_mem dst_image,
...
+            static_cast<cvk_buffer*>(src_img->buffer()), dst_img,
+            src_origin[0] * src_img->element_size(), dst_orig, reg);
+
+        return command_queue->enqueue_command_with_deps(
+            cmd.release(), num_events_in_wait_list, event_wait_list, event);
+    } else if (dst_img->is_backed_by_buffer_view()) {
+        auto cmd = std::make_unique<cvk_command_buffer_image_copy>(
+            CL_COMMAND_COPY_IMAGE_TO_BUFFER, command_queue,
```
Comment (kpet) [inline src/api.cpp:5201]:
```suggestion
            CL_COMMAND_COPY_IMAGE, command_queue,
```

File: src/api.cpp [hunk 9]
```diff
@@ -5168,11 +5265,27 @@ cl_int CLVK_API_CALL clEnqueueFillImage(
...
 
+    auto img = static_cast<cvk_image*>(image);
+
+    // Fill
+    cvk_image::fill_pattern_array pattern;
+    size_t pattern_size;
+    img->prepare_fill_pattern(fill_color, pattern, &pattern_size);
+
+    if (img->is_backed_by_buffer_view()) {
+        auto cmd = new cvk_command_fill_buffer(
```
Comment (kpet) [inline src/api.cpp:5276]: This will expose the wrong command type to applications.

File: src/api.cpp [hunk 10]
```diff
@@ -5279,12 +5387,23 @@ cl_int CLVK_API_CALL clEnqueueCopyImageToBuffer(
                                     src_origin[2]};
     std::array<size_t, 3> reg = {region[0], region[1], region[2]};
 
-    auto cmd = std::make_unique<cvk_command_buffer_image_copy>(
-        CL_COMMAND_COPY_IMAGE_TO_BUFFER, command_queue, buffer, image,
-        dst_offset, origin, reg);
+    if (image->is_backed_by_buffer_view()) {
+        auto cmd = new cvk_command_copy_buffer(
+            command_queue, CL_COMMAND_COPY_BUFFER,
```
Comment (kpet) [inline src/api.cpp:5392]: Wrong command type

File: src/api.cpp [hunk 11]
```diff
@@ -5355,12 +5474,23 @@ cl_int CLVK_API_CALL clEnqueueCopyBufferToImage(
                                     dst_origin[2]};
     std::array<size_t, 3> reg = {region[0], region[1], region[2]};
 
-    auto cmd = std::make_unique<cvk_command_buffer_image_copy>(
-        CL_COMMAND_COPY_BUFFER_TO_IMAGE, command_queue, buffer, image,
-        src_offset, origin, reg);
+    if (image->is_backed_by_buffer_view()) {
+        auto cmd = new cvk_command_copy_buffer(
+            command_queue, CL_COMMAND_COPY_BUFFER, buffer,
```
Comment (kpet) [inline src/api.cpp:5479]: Same here

File: src/api.cpp [hunk 12]
```diff
@@ -5499,10 +5629,20 @@ void* CLVK_API_CALL clEnqueueMapImage(
...
 
     cl_int err;
-    auto ret = cvk_enqueue_map_image(command_queue, image, blocking_map,
-                                     map_flags, origin, region, image_row_pitch,
-                                     image_slice_pitch, num_events_in_wait_list,
-                                     event_wait_list, event, &err);
+    void* ret;
+    auto img = static_cast<cvk_image*>(image);
+    if (img->is_backed_by_buffer_view()) {
+        ret = cvk_enqueue_map_buffer(
```
Comment (kpet) [inline src/api.cpp:5635]: Wrong command type

**Review by kpet — CHANGES_REQUESTED** [review]
This change does not introduce CTS regressions at my end. As discussed I'm not a big fan of handling the buffer view special case in the API layer as this breaks encapsulation. In the interest of getting forward progress, I'm with this approach as a first step (and I think it likely makes the code a bit shorter).

File: src/memory.cpp [hunk 13]
```diff
@@ -485,6 +490,55 @@ bool cvk_image::init() {
...
+
+    auto vkbuf = static_cast<cvk_buffer*>(buffer())->vulkan_buffer();
+
+    VkBufferViewCreateInfo createInfo = {
+        VK_STRUCTURE_TYPE_BUFFER_VIEW_CREATE_INFO,
+        nullptr,
+        0,            // flags
+        vkbuf,        // buffer
+        format,       // format
+        0,            // offset
```
[the remark this answers is not in the document]
Comment (rjodinchr) [reply src/memory.cpp]: I've tried to add something in the CTS ( in `kernel_read_write`), but it just got too complicated and got lost in this code. So I have just added a test for it in clvk.

File: src/api.cpp (hunk 1 above)
Comment (rjodinchr) [reply src/api.cpp]: Done

File: src/api.cpp (hunk 2 above)
Comment (rjodinchr) [reply src/api.cpp:4863]: I am not sure, this is coming from your initial commit I believe. Thinking about it, I don't see a reason to have it, maybe we should get rid of it?

File: src/api.cpp (hunk 3 above)
Comment (rjodinchr) [reply src/api.cpp]: Done

File: src/api.cpp (hunk 4 above)
Comment (rjodinchr) [reply src/api.cpp]: Done

File: src/api.cpp (hunk 6 above)
Comment (rjodinchr) [reply src/api.cpp]: Done

File: src/api.cpp (hunk 7 above)
Comment (rjodinchr) [reply src/api.cpp]: I have added a `cl_command_type` in `cvk_command_buffer_image_copy` so that we have the reported command type as well as the one specifying what copy to perform.

File: src/api.cpp (hunk 8 above)
Comment (rjodinchr) [reply src/api.cpp]: same as above

File: src/api.cpp (hunk 10 above)
Comment (rjodinchr) [reply src/api.cpp]: Fixed

File: src/api.cpp (hunk 11 above)
Comment (rjodinchr) [reply src/api.cpp]: Fixed

File: src/api.cpp (hunk 12 above)
Comment (rjodinchr) [reply src/api.cpp:5617]: Fixed

File: src/api.cpp (hunk 9 above)
Comment (rjodinchr) [reply src/api.cpp:5258]: Fixed

**Comment by kpet** [comment]
@rjodinchr In case you hadn't noticed: it looks like the `1DBufferImageFromSubBuffer` test is hitting an assert.

**Comment by rjodinchr** [comment]
The tests are expected to fail one way or another as this PR is not updating clspv.

**Comment by kpet** [comment]
Ok, makes sense. Can you include the clspv update here (just the one commit we need please) so the CI passes? Thanks!

File: src/api.cpp (hunk 2 above)
Comment (kpet) [reply src/api.cpp:4863]: yeah, if we pass the CTS without it, then we should remove it. If we somehow need this to pass the CTS, we should find a clearer expression of the conditions.

File: src/queue.hpp [hunk 14]
```diff
@@ -957,6 +967,7 @@ struct cvk_command_buffer_image_copy final : public cvk_command_batchable {
     size_t m_offset;
     std::array<size_t, 3> m_origin;
     std::array<size_t, 3> m_region;
+    cl_command_type m_copy_type;
```
Comment (kpet) [inline src/queue.hpp:970]: Reusing `cl_command_type` is a bit confusing but it'll do for a first implementation.

**Review by kpet — COMMENTED** [review]
Almost there. I think we just need to resolve the one open thread about component mapping and update clspv (just the one commit this PR needs) so the CI can pass. Everything else looks nice!

File: src/memory.cpp (hunk 13 above)
[the remark this answers is not in the document]
Comment (kpet) [reply src/memory.cpp]: No problem. The unit test is enough.

**Comment by kpet** [comment]
@rjodinchr I've had a look at the CI failures. The newly added test creates a sub-buffer at an offset that does not satisfy Vulkan constraints. clvk should have rejected the creation of the sub-buffer and texel buffer alignment constraints should be taken into account when calculating `CL_DEVICE_MEM_BASE_ADDR_ALIGN`. I suggest we don't fix sub-buffer error handling in this PR. In the interest of getting forward progress, you could just modify the offsets used in the test so that it works in the CI or query `CL_DEVICE_MEM_BASE_ADDR_ALIGN`, use it to pick the sub-buffer offset and change its implementation to take `storageTexelBufferOffsetAlignmentBytes`, `storageTexelBufferOffsetSingleTexelAlignment`, `uniformTexelBufferOffsetAlignmentBytes`, and `uniformTexelBufferOffsetSingleTexelAlignment` into account.

**Review by kpet — COMMENTED** [review]
It seems swiftshader does not support `VK_FORMAT_FEATURE_2_STORAGE_WRITE_WITHOUT_FORMAT_BIT_KHR` on any format for texel buffers so you're going to have to disable the test for swiftshader.

File: src/api.cpp (hunk 2 above)
Comment (rjodinchr) [reply src/api.cpp:4863]: It seems that this test is needed. Otherwise the CTS thinks that we support `CL_INTENSITY` and `CL_LUMINANCE` with 1d buffer, which does not work with the actual implementation. As they are not mandatory, it feels ok to remove them with this test. What do you think?

Comment (kpet) [reply src/api.cpp:4863]: If they're not supported we should remove them but the condition used to do that is a hack (you said it was part of my draft patch, I'm assuming I'd done that as a quick way of excluding those formats). I've added an entry to #625 so we don't forget to revisit this.

**Comment by kpet** [comment]
Merging this as is to unblock clspv updates. I will fix the CI with a separate commit on top.
